src/zeroD/IdealGasConstPressureMoleReactor.cpp

```diff
@@ -132,92 +132,82 @@ void IdealGasConstPressureMoleReactor::eval(double time, double* LHS, double* RH
     }
 }
 
-Eigen::SparseMatrix<double> IdealGasConstPressureMoleReactor::jacobian(double t,
-    double* y)
+Eigen::SparseMatrix<double> IdealGasConstPressureMoleReactor::jacobian()
 {
+    if (m_nv == 0) {
+        throw CanteraError("IdealGasConstPressureMoleReactor::jacobian",
+                           "Reactor must be initialized first.");
+    }
+
+    vector_fp yCurrent(m_nv);
+    getState(yCurrent.data());
+
     // clear former jacobian elements
     m_jac_trips.clear();
     // Determine Species Derivatives
     // volume / moles * rates portion of equation
-    size_t nspecies = m_thermo->nSpecies();
-    // create sparse structures for rates and volumes
-    Eigen::SparseMatrix<double> rates(nspecies, 1);
-    Eigen::SparseMatrix<double> volumes(1, nspecies);
-    // reserve space for data
-    rates.reserve(nspecies);
-    volumes.reserve(nspecies);
-    // fill sparse structures
-    m_kin->getNetProductionRates(rates.valuePtr()); // "omega dot"
-    std::fill(volumes.valuePtr(), volumes.valuePtr() + nspecies, m_vol);
-    // get ROP derivatives
-    double scalingFactor = m_vol/accumulate(y + m_sidx, y + m_nv, 0.0);
-    Eigen::SparseMatrix<double> speciesDervs = m_kin->netProductionRates_ddX();
-    // sum parts
-    speciesDervs = scalingFactor * speciesDervs + rates * volumes;
-    // add elements to jacobian triplets
-    for (int k=0; k<speciesDervs.outerSize(); ++k) {
-        for (Eigen::SparseMatrix<double>::InnerIterator it(speciesDervs, k); it; ++it) {
-            m_jac_trips.emplace_back(it.row() + m_sidx, it.col() + m_sidx, it.value());
+    Eigen::VectorXd netProductionRates(m_nsp);
+    m_kin->getNetProductionRates(netProductionRates.data()); // "omega dot"
+    Eigen::SparseMatrix<double> dwdX = m_kin->netProductionRates_ddX();
+    double molarVolume = m_thermo->molarVolume();
+    // Calculate ROP derivatives, excluding the term
+    // molarVolume * (wdot(j) - sum_k(X_k * dwdot_j/dX_k))
+    // which is small and would completely destroy the sparsity of the Jacobian
+    for (int k = 0; k < dwdX.outerSize(); k++) {
+        for (Eigen::SparseMatrix<double>::InnerIterator it(dwdX, k); it; ++it) {
+            m_jac_trips.emplace_back(it.row() + m_sidx, it.col() + m_sidx,
+                                     it.value() * molarVolume);
```

Comment:
Using molar volume is a nice addition. I am a bit confused about the exclusion of `molarVolume * (wdot(j) - sum_k(X_k * dwdot_j/dX_k))` in the mole based system as I do not recall `sum_k(X_k * dwdot_j/dX_k)` being involved when everything is converted to moles. The inclusion of the `rates * volumes` terms in the PROCI paper didn't cause an issue. I guess I would include a flag of some sort to exclude these terms if wanted maybe in the `derivative_settings`. Does excluding these terms as causing any kind of issues when integrating?

Reply:
The derivation of the derivatives in the paper is correct, but is given in terms of `dwdot_j/dn_i`, not `dwdot_j/dX_i`. When you use the chain rule on the former to make the latter show up so we can use the derivatives calculated by `GasKinetics`, you get the additional term since `dX_i/dn_j` is not zero even for `i != j`.

The previous implementation for calculating the `rates * volumes` term had a bug in it that caused this matrix to be empty, so it was accidentally being excluded.

Reply:
That makes sense, I believe neglecting those terms is appropriate then. At least until a derivative for `dwdot_j/dn_j` manages to make it into the code.